country codes are missing for some entries.

In [1]:
import csv

##### Task 1

In [2]:
data = list(csv.reader(open('Summer_olympic_Medals.csv','r')))

In [3]:
data[0:5]

[['Year',
  'Host_country',
  'Host_city',
  'Country_Name',
  'Country_Code',
  'Gold',
  'Silver',
  'Bronze'],
 ['1896', 'Greece', 'Athens', 'Great Britain', 'GBR', '2', '3', '2'],
 ['1896', 'Greece', 'Athens', 'Hungary', 'HUN', '2', '1', '3'],
 ['1896', 'Greece', 'Athens', 'France', 'FRA', '5', '4', '2'],
 ['1896', 'Greece', 'Athens', 'United States', 'USA', '11', '7', '2']]

##### Task 2

In [15]:
header = data.pop(0)

In [5]:
header

['Year',
 'Host_country',
 'Host_city',
 'Country_Name',
 'Country_Code',
 'Gold',
 'Silver',
 'Bronze']

In [6]:
len(data)

1344

In [18]:
data[0:5],data[-5:]

([['1896', 'Greece', 'Athens', 'Hungary', 'HUN', '2', '1', '3'],
  ['1896', 'Greece', 'Athens', 'France', 'FRA', '5', '4', '2'],
  ['1896', 'Greece', 'Athens', 'United States', 'USA', '11', '7', '2'],
  ['1896', 'Greece', 'Athens', 'Germany', 'GER', '6', '5', '2'],
  ['1896', 'Greece', 'Athens', 'Austria', 'AUT', '2', '1', '2']],
 [['2020', 'Japan', 'Tokyo', 'Fiji', 'FIJ', '1', '0', '1'],
  ['2020', 'Japan', 'Tokyo', 'Estonia', 'EST', '1', '0', '1'],
  ['2020', 'Japan', 'Tokyo', 'Latvia', 'LAT', '1', '0', '1'],
  ['2020', 'Japan', 'Tokyo', 'Bermuda', 'BER', '1', '0', '0'],
  ['2020', 'Japan', 'Tokyo', 'Thailand', 'THA', '1', '0', '1']])

There are 1344 entries plus a header (column names) in the data.

#### Tasks 3 to 10

Structure for storing countries, years, medals: {country: [[year1, golds1, silvers1, brozes1], ...[yearN, goldsN, silversN, brozesN]]}

In [7]:
data_dict = {}
country_codes = {}

for y_c in data:  
    # loop through the data as unique year and country combinations
    codes_countries = {country:code for code, country in country_codes.items()}
    if y_c[4]=='':
        try:
            y_c[4] = codes_countries[y_c[3]]
        except:
            y_c[4] = "_".join(y_c[3].split(' '))
    if y_c[4] in data_dict:
        data_dict[y_c[4]] = data_dict[y_c[4]] + [[y_c[0], y_c[5], y_c[6], y_c[7]]]
    else: 
        country_codes[y_c[4]]=y_c[3] # correspond country codes with full name
        data_dict[y_c[4]] = [[y_c[0], y_c[5], y_c[6], y_c[7]]]

In [8]:
data_dict = {c: [list(map(int,val)) for val in y_c] for c,y_c in data_dict.items()}

##### Task 3

Instruction: compute the average numbers of gold, silver, and bronze medals per country and write this (in decreasing order) to a file ”average medals per country.csv”

Compute total medals each year

In [9]:
data_totals_dict = {c: [y_c + [sum(y_c[1:])] for y_c in data_dict[c]] for c in data_dict}

In [10]:
avg_medals = {}
avg_medals = {c: list(map(lambda x: x/len(data_totals_dict[c]), # averaged over years
     [sum([y[1] for y in  data_totals_dict[c]]), # total of golds
     sum([y[2] for y in  data_totals_dict[c]]), # silvers
     sum([y[3] for y in  data_totals_dict[c]]),  # bronzes
     sum([y[4] for y in  data_totals_dict[c]])])) # all medals
     for c in data_totals_dict} # averaged

In [11]:
avg_medals_list = [['Country', 'Avg_Gold_Medals','Avg_Silver_Medals',
                'Avg_Bronze_Medals','Avg_Total_Medals']] + \
[[country_codes[c[0]]]+ c[1] for c in list(dict(sorted(
    {c:m for c,m in avg_medals.items()}.items(), 
    key = lambda x: x[1][3], reverse=True)).items())
]

In [12]:
with open('average_medals_per_country.csv','w') as medals_csv:
    wr = csv.writer(medals_csv)
    for row in avg_medals_list:
        wr.writerow(row)
    medals_csv.close()
# medals_csv = open('average_medals_per_country.csv','r')
# medals_csv.read()

##### Highest Average Medals

In [13]:
avg_medals_list[1]

['Soviet Union',
 43.888888888888886,
 35.44444444444444,
 32.888888888888886,
 112.22222222222223]

Highest avg medals: Soviet Union

##### 5. Top 10 countries by mean number of (total) medals

In [14]:
top_ten_avg_medals = [c[0] for c in medals_list[1:11]]

NameError: name 'medals_list' is not defined

In [ ]:
top_ten_avg_medals

['Soviet Union',
 'Unified Team',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'West Germany',
 'United Team of Germany',
 'Germany',
 'Great Britain'] have the highest average medals.

##### 6.  compute the median number of gold, silver, and bronze medals per country and write this (in decreasing order) to a file "median_medals_ per_country.csv"

In [ ]:
def median(numbers):
    '''take a list or tuple of numbers and return the median'''
    size = len(numbers)
    numbers = sorted(numbers)
    if size == 1:
        med = numbers[0]
    elif size%2 == 0:
        med = (numbers[int(size/2-1)]+numbers[int(size/2)])/2
    else:
        med = numbers[size//2]
    return med

In [ ]:
med_medals_dict = {c:[median([y[1] for y in data_totals_dict[c]])] + \
                   [median([y[2] for y in data_totals_dict[c]])] + \
                   [median([y[3] for y in data_totals_dict[c]])] + \
                   [median([y[4] for y in data_totals_dict[c]])]
                   for c in data_totals_dict.keys()}

In [ ]:
median_medals = dict(sorted(med_medals_dict.items(), key = lambda x: x[1][3], reverse=True))
median_medals

In [ ]:
med_medals_list = [['Country', 'Med_Gold_Medals','Med_Silver_Medals',
                'Med_Bronze_Medals','Med_Total_Medals']] + \
[[country_codes[c[0]]]+ c[1] for c in list(dict(sorted(
    {c:m for c,m in median_medals.items()}.items(), 
    key = lambda x: x[1][3], reverse=True)).items())
]

In [ ]:
with open('median_medals_per_country.csv','w') as median_medals_csv:
    wr = csv.writer(median_medals_csv)
    for row in med_medals_list:
        wr.writerow(row)
    median_medals_csv.close()
# median_medals_csv = open('median_medals_per_country.csv','r')
# median_medals_csv.read()

##### 7. Which country has the highest median number of medals?

In [ ]:
med_medals_list[1][0]

Q7 Answer: Unified Team has the highest median number of medals.

##### 8. list top 10 countries by median number of medals

In [ ]:
[c[0] for c in med_medals_list[1:11]]

Q8 answer: 'Unified Team',
 'Soviet Union',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'United Team of Germany',
 'West Germany',
 'Germany',
 'Great Britain'

##### 9. For which countries you have #(gold) = #(silver) = #(bronze). For which countries you have (#(gold) + #(silver) ) = #(bronze)

###### 9a: which countries have the same numbers of gold, silver, and bronze medals.

In [ ]:
# make a dictionary that corresponds the country with their aggregated totals for 
# gold, silver, bronze, and overall accross all years

medal_totals = {c:  # averaged over years
     [sum([y[1] for y in  data_totals_dict[c]]), # total of golds
     sum([y[2] for y in  data_totals_dict[c]]), # silvers
     sum([y[3] for y in  data_totals_dict[c]]),  # bronzes
     sum([y[4] for y in  data_totals_dict[c]])] # all medals
     for c in data_totals_dict} # averaged

In [ ]:
{c:m for c,m in medal_totals.items() if (m[0]==m[1])&(m[1]==m[2])}

In [ ]:
[country_codes[c] for c in {c:m for c,m in medal_totals.items() if (m[0]==m[1])&(m[1]==m[2])}.keys()]

9a: Trinidad and Tobago along with Grenada are the two countries that have equal numbers of gold, silver and bronze medals. 

###### 9b: Which countries have the same number of bronze medals as the sum of gold and silver medals.

In [ ]:
[country_codes[c] for c in 
 {c:m for c,m in medal_totals.items() if (m[0]+m[1])==m[2]}]

9b Answer: 'Haiti',
 'Portugal',
 'Egypt',
 'Lebanon',
 'Iceland',
 'Morocco',
 'Niger',
 'Bermuda',
 'Zambia',
 'Syria',
 'Ivory Coast',
 'Chinese Taipei',
 'Suriname',
 'Costa Rica',
 'Lithuania',
 'Kazakhstan',
 'Mozambique',
 'Saudi Arabia',
 'Macedonia',
 'United Arab Emirates',
 'Tajikistan',
 'Botswana',
 'Independent Olympic Athletes',
 'Jordan',
 'Fiji'

##### 10. Which countries increased their number of total medals each game during the last 4 Olympic games and which coun- tries decreased their total number of medals during the last 4 Olympic games?

In [ ]:
# Determine who participated in the last 5 Olympics
def participated_l_5(d = dict()):
    '''take a dictionary of coutry:[list of years and medals] and determine if 
    they competed in all of the last five Olympics, 
    (2000, 2004, 2008, 2012, and 2020), returning a revised dictionary with countries
    remaining if they satisfy this condition.'''
    participated = {}
    for country, year_medals in d.items():
        if [y[0] for y in year_medals[-5:]] == \
        [2004, 2008, 2012, 2016, 2020]:
            participated[country] = True
        else:
            participated[country] = False
    return {c:d[c] for c,p in participated.items() if p == True}

###### 10a Increasing medals, i.e. 2020<2016<2012<2008<2004.

In [ ]:
def increasing_medals(d = dict()):
    '''return a dictionary including the items for which there is an increase
    in total medals for each of the last four olympics'''
    return {c: y_m for c, y_m in d.items() if \
                 (y_m[-1][4] > y_m[-2][4])&\
                 (y_m[-2][4] > y_m[-3][4])&\
                 (y_m[-3][4] > y_m[-4][4])\
                 &(y_m[-4][4] > y_m[-5][4])
           }

In [ ]:
participated_l_5(data_totals_dict)

In [ ]:
[country_codes[code] for code in 
 increasing_medals(participated_l_5(data_totals_dict)).keys()]

In [ ]:
increasing_medals(participated_l_5(data_totals_dict))

New Zealand is the only country that increased its medal count over the last four olympics, going up from 5 in 2004 to 9, 13, and a finally a very impressive 20 in 2020.  Belgium and Japan both increased their medal counts over the last three Olympics.

###### 10b Decreased over last four olympics

I am intrepreting this to mean 2020<2016<2012<2008<2004.

In [ ]:
def decreasing_medals(d = dict()):
    '''return a dictionary including the items for which there is an increase
    in total medals for each of the last four olympics'''
    return {c: y_m for c, y_m in d.items() if \
     (y_m[-1][4] < y_m[-2][4])&\
     (y_m[-2][4] < y_m[-3][4])&\
     (y_m[-3][4] < y_m[-4][4])\
     &(y_m[-4][4] < y_m[-5][4])
    }

In [ ]:
[country_codes[code] for code in 
 decreasing_medals(participated_l_5(data_totals_dict)).keys()]

No country decreased its medals count over each of the last four Olympics. South Korea and Belarus decreased their medal counts over the last three Olympics.